robots/pkg/circuitbreaker/circuitbreaker.go

```diff
@@ -63,17 +67,17 @@ func (g *CircuitBreaker) WrapRetryableFunc(retryableFunc func() error) func() er
 }
 
 func (g *CircuitBreaker) isOpenAndNotFeasibleForRetry() (bool, error) {
-	g.mutex.Lock()
-	defer g.mutex.Unlock()
-	return g.open && !g.occurred.Add(g.retryAfter).Before(time.Now()), g.lastErr
+	g.mutex.RLock()
+	defer g.mutex.RUnlock()
+	return g.open && !g.blockedUntil.Before(time.Now()), g.lastErr
 }
 
 func (g *CircuitBreaker) updateState(err error) {
 	g.mutex.Lock()
 	defer g.mutex.Unlock()
-	if err != nil {
+	if err != nil && g.shouldOpen(err) {
```

Comment:
Sorry for not noticing this earlier, but here probably we should check again if the circuit is open.
It should not happen, but let me try to explain my concern:
- A makes a request X at t<sup>0</sup>.
- B makes a request Y at t<sup>1</sup>.

Both A and B see the closed circuit, so the requests X,Y are executed.
Now let's assume that after a while(t<sup>3</sup>) X return 504; the circuit will be open with `blockedUntil` t<sup>5</sup>.
But after another while(t<sup>4</sup>) Y return 504 and the `blockedUntil` will be updated with t<sup>6</sup>.
Taking the reasoning to extremes, it may happen that the circuit remains open for twice the time it was set.
That is, the case where a 504 response to a request arrives exactly an instant before `blockedUntil`, as it will override it just a moment before it expires.
Checking the circuit before the write ensures that `blockedUntil` always refer to the first failure.
Sorry for the probably unclear comment. Let me know. Thanks

Reply:
What you are describing to happen should be fine, since the failure of the second call indicates that the service under supervision is still not operating as it should, in which case the circuit breaker should extend the wait time for the latter occurrence. WDYT?

Reply:
yes, but the service could be available after a `retryAfter` amount of time from the first occurrence of the error, but since we've updated the `blockUntil` just before the end of the first timeout we are going to wait for another `retryAfter` amount of time.

Reply:
Added a check at the beginning of the update.